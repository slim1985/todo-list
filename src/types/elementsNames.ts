export enum ElementsNames {
    STATUS_SELECT = 'status-select',
    TITLE_TEXT_AREA = 'title-textarea',
    DESCRIPTION_TEXT_AREA = 'description-textarea',
}
