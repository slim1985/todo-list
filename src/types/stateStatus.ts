export enum StateStatus {
    LOADING = 'loading',
    IDLE = 'idle',
    FAILED = 'failed',
}
