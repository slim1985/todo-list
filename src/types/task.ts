export interface Task {
    id: string;
    title: string;
    description: string;
    state: TaskStates;
}

export enum TaskStates {
    new = 'new',
    active = 'active',
    completed = 'completed',
}
