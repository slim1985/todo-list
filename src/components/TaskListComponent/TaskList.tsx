import { useState } from 'react';
import { Task, TaskStates } from '../../types/task';
import { TaskCardMemo } from '../TaskCardComponent/TaskCard';
import { TaskForm } from '../TaskFormComponent/TaskForm';

export interface TaskListPanelProps {
    taskList: Task[];
    saveTask: (
        id: string,
        title: string,
        description: string,
        status: TaskStates,
    ) => void;
    deleteTask: (taskId: string) => void;
}

export function TaskList({
    taskList,
    saveTask,
    deleteTask,
}: TaskListPanelProps): JSX.Element {
    const [showTaskForm, setShowTaskForm] = useState(false);
    const [selectedTask, setSelectedTask] = useState<Task>();

    function openTaskDetails(id: string): void {
        const task = taskList.find((task) => task.id === id)!;
        setSelectedTask(task);
        setShowTaskForm(true);
    }

    function createTask(): void {
        const newTask: Task = {
            id: '',
            title: 'Title',
            description: 'Description',
            status: TaskStates.NEW,
        };

        setSelectedTask(newTask);
        setShowTaskForm(true);
    }

    return (
        <>
            <div>
                <div className="flex justify-end sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2">
                    <button
                        onClick={() => createTask()}
                        className="size-20 my-2 mr-2 p-1 bg-gray-300 rounded-md"
                    >
                        Create
                    </button>
                </div>
                <div className="flex flex-wrap justify-center mt-3">
                    {taskList.map((task: Task, index: number) => (
                        <TaskCardMemo
                            key={index}
                            task={task}
                            onClick={openTaskDetails}
                        />
                    ))}
                </div>
            </div>
            {showTaskForm && (
                <TaskForm
                    task={selectedTask!}
                    setShowTaskForm={setShowTaskForm}
                    saveTask={saveTask}
                    deleteTask={deleteTask}
                />
            )}
        </>
    );
}
