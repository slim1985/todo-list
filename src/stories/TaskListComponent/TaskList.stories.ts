import type { Meta, StoryObj } from '@storybook/react';
import { TaskList } from '../../components/TaskListComponent/TaskList';
import { Task } from '../../types/task';
import tasks from '../../mocks/data-mock.json';

type Story = StoryObj<typeof meta>;

const meta: Meta<typeof TaskList> = {
    component: TaskList,
    parameters: {
        viewport: {
            defaultViewport: 'mobile1',
        },
    },
};

const taskList = JSON.parse(JSON.stringify(tasks)) as Task[];

export default meta;

export const Simple: Story = {
    name: 'TaskListPanel',
    args: {
        taskList: taskList,
    },
};
