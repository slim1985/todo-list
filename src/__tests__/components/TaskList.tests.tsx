import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { TaskList } from '../../components/TaskListComponent/TaskList';
import { Task, TaskStates, TaskStateLabels } from '../../types/task';

const taskList: Task[] = [
    {
        id: '1',
        title: 'Task 1',
        description: 'Task 1 description',
        status: TaskStates.NEW,
    },
    {
        id: '2',
        title: 'Task 2',
        description: 'Task 2 description',
        status: TaskStates.ACTIVE,
    },
    {
        id: '3',
        title: 'Task 3',
        description: 'Task 3 description',
        status: TaskStates.COMPLETED,
    },
];

const openTask = jest.fn();

describe('TaskList', () => {
    it('TaskList renders correctly', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskList taskList={taskList} openTask={openTask} />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskList data renders correctly', () => {
        // Arrange.
        // Act.
        render(<TaskList taskList={taskList} openTask={openTask} />);

        // Assert.
        taskList.forEach((task) => {
            expect(screen.getByText(task.title)).toBeInTheDocument();
            expect(screen.getByText(task.description)).toBeInTheDocument();
            expect(
                screen.getByText(TaskStateLabels[task.status]),
            ).toBeInTheDocument();
        });
    });

    it('TaskList openTask event is called', () => {
        // Arrange.
        // Act.
        render(<TaskList taskList={taskList} openTask={openTask} />);
        screen.getByText(taskList[0].title).click();

        // Assert.
        expect(openTask).toHaveBeenCalledTimes(1);
    });
});
