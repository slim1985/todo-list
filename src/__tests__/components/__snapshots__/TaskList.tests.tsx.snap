// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList TaskList renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap justify-center mt-3"
  >
    <div
      class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
    >
      <h2
        class="w-full font-bold retina:text-6xl"
      >
        Task 1
      </h2>
      <p
        class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
      >
        Task 1 description
      </p>
      <p
        class="w-full font-bold retina:text-6xl"
      >
        Task state: 
        <span
          class="text-green-700"
        >
          New
        </span>
      </p>
    </div>
    <div
      class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
    >
      <h2
        class="w-full font-bold retina:text-6xl"
      >
        Task 2
      </h2>
      <p
        class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
      >
        Task 2 description
      </p>
      <p
        class="w-full font-bold retina:text-6xl"
      >
        Task state: 
        <span
          class="text-blue-700"
        >
          Active
        </span>
      </p>
    </div>
    <div
      class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
    >
      <h2
        class="w-full font-bold retina:text-6xl"
      >
        Task 3
      </h2>
      <p
        class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
      >
        Task 3 description
      </p>
      <p
        class="w-full font-bold retina:text-6xl"
      >
        Task state: 
        <span
          class="text-gray-700"
        >
          Completed
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
